### ISSUE TITLE
Wrong metadata of tensorflow-cpu-aws in TensorFlow 2.16.1

### ISSUE THREAD

**Issue by tachikoma-li** [tensorflow/tensorflow#63585]
### Issue type

Build/Install

### Have you reproduced the bug with TensorFlow Nightly?

No

### TensorFlow version

2.16.1

### OS platform and distribution

Linux x86_64

### Python version

3.11

### Current behavior?

The metadata of TensoFlow 2.16.1 (for Linux x86_64) has a dependency of `tensorflow-cpu-aws`:
```
Requires-Dist: tensorflow-cpu-aws ==2.16.1 ; platform_system == "Linux" and (platform_machine == "arm64" or platform_machine == "aarch64")
```
However, the latest version of `tensorflow-cpu-aws` is [2.15.0](https://pypi.org/project/tensorflow-cpu-aws/). This would cause trouble when resolving the dependencies with the package metadata.
If we can't ensure the release of `tensorflow-cpu-aws` is synchronized with `tensorflow`, can we relax the dependency requirement in the metadata?

**Comment by juhoinkinen** [comment]
Any updates? This issue makes it impossible to install tensorflow-cpu (any version) on Python  **3.12** when using Poetry. Or is there a workaround?

For example, when using this `pyproject.toml`:
```
[tool.poetry]
name = "python-3-12-poetry-env"
version = "0.1.0"
description = ""
authors = [""]

[tool.poetry.dependencies]
python = ">=3.9,<3.13"
tensorflow-cpu = { version = "2.16.*" }

[build-system]
requires = ["poetry-core"]
build-backend = "poetry.core.masonry.api"
```

And then running `poetry install` the output is
```
Because no versions of tensorflow-cpu match >=2.16.dev0,<2.16.0rc0 || >2.16.0rc0,<2.16.1 || >2.16.1,<2.17.dev0
 and tensorflow-cpu (2.16.0rc0) depends on tensorflow-cpu-aws (2.16.0-rc0), tensorflow-cpu (==2.16.0.* || >2.16.1,<2.17.dev0) requires tensorflow-cpu-aws (2.16.0-rc0).
And because no versions of tensorflow-cpu-aws match 2.16.0-rc0
 and tensorflow-cpu (2.16.1) depends on tensorflow-cpu-aws (2.16.1), tensorflow-cpu (==2.16.*) requires tensorflow-cpu-aws (2.16.1).
So, because no versions of tensorflow-cpu-aws match 2.16.1
 and python-3-12-poetry-env depends on tensorflow-cpu (2.16.*), version solving failed.
```

**Comment by BlackHC** [comment]
This also affects other Python versions, not just 3.12. I'm on 3.10. How am I supposed to depend on this version of TensorFlow when using poetry?

This seems to be broken?!?

**Comment by juhoinkinen** [comment]
Installing tf-nightly-cpu version 2.17.0.dev20240412 with Poetry works, so this is fixed in master (intentionally?). If patch release 2.16.2 is going to be made, the fix of this issue should be included in it.

I tried to identify the fix, but could not understand the publish/build process.

**Comment by nitins17** [comment]
Thanks for letting us know. 

https://github.com/tensorflow/tensorflow/commit/a03c28b65b4491a1c657e5ef0a918913c5db42ec did not make it into r2.16 which is why it is fixed on Master but not in r2.16. https://github.com/tensorflow/tensorflow/pull/67772 should fix it for TF 2.16.2.

In the meantime, I'll take look at the Poetry installation to see if something can done for mitigation until then.

**Comment by nitins17** [comment]
**TL;DR:** Poetry encounters dependency resolution issues with `tensorflow-cpu` due to a dependency on `tensorflow-cpu-aws`. Fixed wheels for TF 2.16.1 are available in a GCS bucket with instructions on how to use them. This is a temporary workaround until TF 2.16.2 is released.
***
I created a sample project using Poetry to investigate this issue. I found that Poetry successfully resolves dependencies when using `tensorflow`, but encounters issues with `tensorflow-cpu`.

Specifically, the following configuration works:
```
[tool.poetry.dependencies]
python = "^3.9"
tensorflow = "^2.16.1"
```

While this configuration does not:
```
[tool.poetry.dependencies]
python = "^3.9"
tensorflow_cpu = "^2.16.1"
````

As others have noted, the problem arises when resolving the dependency on `tensorflow-cpu-aws`:
```
Because no versions of tensorflow-cpu match >2.16.1,<3.0.0
 and tensorflow-cpu (2.16.1) depends on tensorflow-cpu-aws (2.16.1), tensorflow-cpu (>=2.16.1,<3.0.0) requires tensorflow-cpu-aws (2.16.1).
So, because no versions of tensorflow-cpu-aws match 2.16.1
 and test-project depends on tensorflow_cpu (^2.16.1), version solving failed.
 ````
 
Upon inspecting the metadata of both `tensorflow` and `tensorflow-cpu` wheel packages, I noticed they are nearly identical except for the package name. Interestingly, both contain the following requirement:

 ```
 Requires-Dist: tensorflow-cpu-aws ==2.16.1 ; platform_system == "Linux" and (platform_machine == "arm64" or platform_machine == "aarch64")
```

It's unclear why depending on `tensorflow` works, while `tensorflow-cpu` triggers an error.  In any case, the dependency on `tensorflow-cpu-aws` in `tensorflow_cpu-2.16.1.dist-info/METADATA` seems to be the culprit.  Removing this requirement allows Poetry to successfully resolve the `tensorflow-cpu` dependency.

I've applied this fix to all `tensorflow` and `tensorflow-cpu` wheels for TF 2.16.1, available in `gs://tensorflow-2-16-1-fixed-wheels-poetry` (see end of this comment that lists all the wheels). To use them, modify your `pyproject.toml` as following:
```
tensorflow-cpu = {url="https://storage.googleapis.com/tensorflow-2-16-1-fixed-wheels-poetry/tensorflow_cpu/tensorflow_cpu-2.16.1-cp311-cp311-manylinux_2_17_x86_64.manylinux2014_x86_64.whl" }
```

@randolf-scholz While I haven't tested it, it seems that this approach should also work with pdm, based on their [documentation](https://pdm-project.org/en/latest/usage/dependency/#url-dependencies).

I hope this temporary solution helps until the release of TF 2.16.2 (expected in the next month or two). Please let me know if anyone encounter any further issues.

<details>
<summary> tensorflow </summary>
<pre>
<code>
$ gsutil ls gs://tensorflow-2-16-1-fixed-wheels-poetry/tensorflow/*.whl
gs://tensorflow-2-16-1-fixed-wheels-poetry/tensorflow/tensorflow-2.16.1-cp310-cp310-macosx_10_15_x86_64.whl
gs://tensorflow-2-16-1-fixed-wheels-poetry/tensorflow/tensorflow-2.16.1-cp310-cp310-macosx_12_0_arm64.whl
gs://tensorflow-2-16-1-fixed-wheels-poetry/tensorflow/tensorflow-2.16.1-cp310-cp310-manylinux_2_17_aarch64.manylinux2014_aarch64.whl
gs://tensorflow-2-16-1-fixed-wheels-poetry/tensorflow/tensorflow-2.16.1-cp310-cp310-manylinux_2_17_x86_64.manylinux2014_x86_64.whl
gs://tensorflow-2-16-1-fixed-wheels-poetry/tensorflow/tensorflow-2.16.1-cp310-cp310-win_amd64.whl
gs://tensorflow-2-16-1-fixed-wheels-poetry/tensorflow/tensorflow-2.16.1-cp311-cp311-macosx_10_15_x86_64.whl
gs://tensorflow-2-16-1-fixed-wheels-poetry/tensorflow/tensorflow-2.16.1-cp311-cp311-macosx_12_0_arm64.whl
gs://tensorflow-2-16-1-fixed-wheels-poetry/tensorflow/tensorflow-2.16.1-cp311-cp311-manylinux_2_17_aarch64.manylinux2014_aarch64.whl
gs://tensorflow-2-16-1-fixed-wheels-poetry/tensorflow/tensorflow-2.16.1-cp311-cp311-manylinux_2_17_x86_64.manylinux2014_x86_64.whl
gs://tensorflow-2-16-1-fixed-wheels-poetry/tensorflow/tensorflow-2.16.1-cp311-cp311-win_amd64.whl
gs://tensorflow-2-16-1-fixed-wheels-poetry/tensorflow/tensorflow-2.16.1-cp312-cp312-macosx_10_15_x86_64.whl
gs://tensorflow-2-16-1-fixed-wheels-poetry/tensorflow/tensorflow-2.16.1-cp312-cp312-macosx_12_0_arm64.whl
gs://tensorflow-2-16-1-fixed-wheels-poetry/tensorflow/tensorflow-2.16.1-cp312-cp312-manylinux_2_17_aarch64.manylinux2014_aarch64.whl
gs://tensorflow-2-16-1-fixed-wheels-poetry/tensorflow/tensorflow-2.16.1-cp312-cp312-manylinux_2_17_x86_64.manylinux2014_x86_64.whl
gs://tensorflow-2-16-1-fixed-wheels-poetry/tensorflow/tensorflow-2.16.1-cp312-cp312-win_amd64.whl
gs://tensorflow-2-16-1-fixed-wheels-poetry/tensorflow/tensorflow-2.16.1-cp39-cp39-macosx_10_15_x86_64.whl
gs://tensorflow-2-16-1-fixed-wheels-poetry/tensorflow/tensorflow-2.16.1-cp39-cp39-macosx_12_0_arm64.whl
gs://tensorflow-2-16-1-fixed-wheels-poetry/tensorflow/tensorflow-2.16.1-cp39-cp39-manylinux_2_17_aarch64.manylinux2014_aarch64.whl
gs://tensorflow-2-16-1-fixed-wheels-poetry/tensorflow/tensorflow-2.16.1-cp39-cp39-manylinux_2_17_x86_64.manylinux2014_x86_64.whl
gs://tensorflow-2-16-1-fixed-wheels-poetry/tensorflow/tensorflow-2.16.1-cp39-cp39-win_amd64.whl
</code>
</pre>
</details>

<details>
<summary> tensorflow-cpu </summary>
<pre>
<code>
$ gsutil ls gs://tensorflow-2-16-1-fixed-wheels-poetry/tensorflow_cpu/*.whl
gs://tensorflow-2-16-1-fixed-wheels-poetry/tensorflow_cpu/tensorflow_cpu-2.16.1-cp310-cp310-macosx_10_15_x86_64.whl
gs://tensorflow-2-16-1-fixed-wheels-poetry/tensorflow_cpu/tensorflow_cpu-2.16.1-cp310-cp310-manylinux_2_17_x86_64.manylinux2014_x86_64.whl
gs://tensorflow-2-16-1-fixed-wheels-poetry/tensorflow_cpu/tensorflow_cpu-2.16.1-cp310-cp310-win_amd64.whl
gs://tensorflow-2-16-1-fixed-wheels-poetry/tensorflow_cpu/tensorflow_cpu-2.16.1-cp311-cp311-macosx_10_15_x86_64.whl
gs://tensorflow-2-16-1-fixed-wheels-poetry/tensorflow_cpu/tensorflow_cpu-2.16.1-cp311-cp311-manylinux_2_17_x86_64.manylinux2014_x86_64.whl
gs://tensorflow-2-16-1-fixed-wheels-poetry/tensorflow_cpu/tensorflow_cpu-2.16.1-cp311-cp311-win_amd64.whl
gs://tensorflow-2-16-1-fixed-wheels-poetry/tensorflow_cpu/tensorflow_cpu-2.16.1-cp312-cp312-macosx_10_15_x86_64.whl
gs://tensorflow-2-16-1-fixed-wheels-poetry/tensorflow_cpu/tensorflow_cpu-2.16.1-cp312-cp312-manylinux_2_17_x86_64.manylinux2014_x86_64.whl
gs://tensorflow-2-16-1-fixed-wheels-poetry/tensorflow_cpu/tensorflow_cpu-2.16.1-cp312-cp312-win_amd64.whl
gs://tensorflow-2-16-1-fixed-wheels-poetry/tensorflow_cpu/tensorflow_cpu-2.16.1-cp39-cp39-macosx_10_15_x86_64.whl
gs://tensorflow-2-16-1-fixed-wheels-poetry/tensorflow_cpu/tensorflow_cpu-2.16.1-cp39-cp39-manylinux_2_17_x86_64.manylinux2014_x86_64.whl
gs://tensorflow-2-16-1-fixed-wheels-poetry/tensorflow_cpu/tensorflow_cpu-2.16.1-cp39-cp39-win_amd64.whl
</code>
</pre>
</details>

**Comment by tachikoma-li** [comment]
@nitins17, thanks for addressing this. Unfortunately, this would not work for us as we are not using poetry and the patched wheel file is for 3.11 only. Just wondering when can we expect the release of 2.16.2? This is blocking us from supporting Python 3.12 as we can't use 2.16.* in our dependencies.

**Comment by nitins17** [comment]
@tachikoma-li Patched versions of all supported wheels should be present in `gs://tensorflow-2-16-1-fixed-wheels-poetry`.  Sorry if it wasn't clear in my earlier post. In order to download a specific wheel, just replace `<wheel_name>` in the following urls:
1. `tensorflow`: `https://storage.googleapis.com/tensorflow-2-16-1-fixed-wheels-poetry/tensorflow/<wheel_name>`
2. `tensorflow-cpu`: `https://storage.googleapis.com/tensorflow-2-16-1-fixed-wheels-poetry/tensorflow_cpu/<wheel_name>`

E.g: If you want to install the patched Python 3.12 wheel of TF 2.16 for Linux x86 then run the following:
```
wget https://storage.googleapis.com/tensorflow-2-16-1-fixed-wheels-poetry/tensorflow/tensorflow-2.16.1-cp312-cp312-manylinux_2_17_x86_64.manylinux2014_x86_64.whl
pip install --no-cache-dir tensorflow-2.16.1-cp312-cp312-manylinux_2_17_x86_64.manylinux2014_x86_64.whl
```

Sadly, I know this isn't the cleanest of solutions but hope it helps until TF 2.16.2 comes out.

> Just wondering when can we expect the release of 2.16.2? 

Branch cut was done this week. I'm guessing here but I think rc0 should come out in the next week or so and then 3-4 weeks more for the final release provided everything goes smoothly.

**Comment by nitins17** [comment]
TF 2.16.2 has been [released](https://pypi.org/project/tensorflow/2.16.2/#files)!

Closing this now as it should be fixed in 2.16.2. Please feel free to reopen if anyone is still running into this issue after upgrading to TF 2.16.2

> I have the same problem with tensorflow-macos

@miljkovicm15 For Apple Silicon wheels, please switch your workflow to install `tensorflow`  directly via https://pypi.org/project/tensorflow.
